treadmill/infra/instances.py

```diff
@@ -0,0 +1,234 @@
+        self.id = id
+        self.name = Name
+        self.metadata = metadata
+        self.conn = Connection()
+
+        if self.metadata and self.metadata.get('Tags', None):
+            self.name = [t['Value']
+                         for t in self.metadata['Tags']
+                         if t['Key'] == 'Name'][0]
+
+        self.private_ip = self.metadata.get(
+            'PrivateIpAddress',
+            ''
+        ) if metadata else ''
+
+    def create_tags(self):
+        self.name = self.name + str(
+            self.metadata.get('AmiLaunchIndex', 0) + 1
+        )
+        self.conn.create_tags(
+            Resources=[self.id],
+            Tags=[{
+                'Key': 'Name',
+                'Value': self.name
+            }]
+        )
+
+    def upsert_dns_record(self, hosted_zone_id, domain='', Reverse=False):
+        self._change_resource_record_sets(
+            'UPSERT',
+            hosted_zone_id,
+            domain,
+            Reverse
+        )
+
+    def delete_dns_record(self, hosted_zone_id, domain='', Reverse=False):
+        self._change_resource_record_sets(
+            'DELETE',
+            hosted_zone_id,
+            domain,
+            Reverse
+        )
+
+    def _change_resource_record_sets(
+            self,
+            action,
+            hosted_zone_id,
+            domain='',
+            Reverse=False
+    ):
+        forward_dns_name = self.name.lower() + '.' + domain + '.'
+        _hosted_zone_id = hosted_zone_id.split('/')[-1]
+        _name, _type, _value = [
+            self._reverse_dns_record_name(),
+            'PTR',
+            forward_dns_name
+        ] if Reverse else [
+            forward_dns_name,
+            'A',
+            self.private_ip
+        ]
+        _conn = Connection('route53')
+        _conn.change_resource_record_sets(
+            HostedZoneId=_hosted_zone_id,
+            ChangeBatch={
+                'Changes': [{
+                    'Action': action,
+                    'ResourceRecordSet': {
+                        'Name': _name,
+                        'Type': _type,
+                        'TTL': 3600,
+                        'ResourceRecords': [{
+                            'Value': _value
+                        }]
+                    }
+                }]
+            }
+        )
+
+    def _reverse_dns_record_name(self):
+        ip_octets = self.private_ip.split('.')
+        ip_octets.reverse()
+        ip_octets.append('in-addr.arpa')
+
+        return '.'.join(ip_octets)
+
+
+class Instances:
+    def __init__(self, instances):
+        self.instances = instances
+        self.volume_ids = []
+        self.conn = Connection()
+
+    @property
+    def ids(self):
+        return [i.id for i in self.instances]
+
+    @classmethod
+    def load_json(cls, ids=[], filters=[]):
+        """Fetch instance details"""
+        conn = Connection()
+        response = []
+
+        if ids:
```

Comment:
If both ids and filters are passed, it will always prefer ids which is compliant with boto3 API as well.
Refactored the part when nothing is specified.